HIVE-25734: Wrongly-typed constant in case expression leads to incorrect empty result

asolimando opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
1) `HivePointLookupOptimizerRule` should not attempt at merging `IN` clauses when the constant types is different, we fix this in the current PR.

2) Another issue emerging in the added qtest query is that constants in `case when` expressions are not typed following the type of the predicate they will be compared too. In the qtest this leads to the creation of `IN` clauses with constants of type `INTEGER` where `SMALLINT` type would have been more appropriate, which makes the optimizer missing opportunities.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
1) Merging `IN` clauses with incompatible types can lead to empty results while some are expected.

2) The second issue does not harm correctness but limits rewrite opportunities.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Queries involving `case when` might have better typing for constants and this can lead to further optimizations, but no user-facing changes apart from correcting the bug.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
The first issue is covered by additional unit tests in `TestHivePointLookupOptimizerRule`, the second one by a new qtest and a new unit test class `TestRexNodeConverter`.

kgyrtkirk commented on ql/src/test/queries/clientpositive/cbo_case_when_wrong_type.q:
```diff
@@ -0,0 +1,20 @@
+create external table test_case (row_seq smallint, row_desc string) stored as parquet;
```
can we also add a simpler testcase like:
```
explain select 1 from t where a in (1,2,3) and case a when 1S then true when 2S then true end;
```

asolimando replied:
Sure, would you prefer a separate qtest file or to integrate it into this one?

asolimando commented:
There were 2 test failures:
- testCliDriver[query39] – org.apache.hadoop.hive.cli.TestTezTPCDS30TBPerfCliDriver (updated but the change looks good, it's exactly our use case)
- testCliDriver[udf_unix_timestamp] – org.apache.hadoop.hive.cli.TestBeeLineDriver (unrelated, passes locally, it proved flacky for me in the past this test)

On top of the updated `.q.output` for `query39` I have added `TestRexNodeConverter` class for unit tests (I have added some corner cases with null/nullability and all seems in order.

I had first expanded the `qtest` file with the second variant of `case when` expression (`case when a = 1` instead of `case a when 1`, where the former was already working as expected), but I have finally simplified the test as you suggested.

If the tests pass and you don't have more comments I will squash the commits and upgrade the commit message.

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HivePointLookupOptimizerRule.java:
```diff
@@ -655,6 +665,26 @@ private static RexNode handleAND(RexBuilder rexBuilder, RexCall call) {
+     * @return true if nodes in both collections is unique and identical, false otherwise
+     */
+    private static boolean shareSameType(Collection<RexNode> nodes1, Collection<RexNode> nodes2) {
+      Set<SqlTypeName> types1 = nodes1.stream()
```
Can any one of this work-

```
     return Stream.of(nodes1, nodes2).flatMap(Collection::stream).map(n -> n.getType().getSqlTypeName()).distinct()
       .count() == 1;
```

```
   return CollectionUtils.union(nodes1, nodes2).stream().map(n -> n.getType().getSqlTypeName()).distinct().count()
          == 1;
```

```     
 return Stream.concat(nodes1.stream(),nodes2.stream()).map(n-> n.getType().getSqlTypeName()).distinct().count() == 1;
```

asolimando replied:
Thanks @ayushtkn, the first one is really what I was looking for, I will adopt it!

asolimando commented:
Test failures seem totally unrelated but not flaky, those 11 tests failed in the two last runs, how do you usually handle such cases?

kgyrtkirk commented:
strange...iirc I've seen this test fail randomly before - I've restarted tests on this PR and launched http://ci.hive.apache.org/job/hive-flaky-check/480/ to check `TestStreamingDynamicPartitioning`

kgyrtkirk commented:
note: if the PR name starts with the associated jira issue id - then a link is added to the jira

asolimando commented:
Tests passed in the last run, can I squash the commits and rephrase the commit message?

kgyrtkirk commented:
no need; we will squash it before merging the changes - if you force-push the branch all the earlier comment will loose context; and the review system also looses the marking of which files are marked as viewed
